### PR TITLE
Add support for parsing `Any` messages in google protobuf text format

### PR DESCRIPTION
**Pull request by tathougies**
This PR implements a special `Registry` object (thanks @judah) that contains a mapping between type URIs and MessageDescriptors. We can create instances of these data types using the `register` function and combine them monadically. The PR implements `showMessageWithRegistry` and `readMessageWithRegistry` which allows the text format parser to read text-formatted messages containing `Any` fields. The `readMessageWithRegistry` function will read and pack such fields into the standard `Any` object. If the type URI can't be found in the registry, an error is emitted. The `showMessageWithRegistry` function uses the registry to encode messages with embedded `Any`s using the more human-readable `Any` syntax. If it encounters an `Any` whose type URI cannot be found in the registry, the `Any` is printed in the packed format.

### CODE REVIEW

**Comment by tathougies** [comment]
I signed it!

File: proto-lens/src/Data/ProtoLens/Message.hs [hunk 1]
```diff
@@ -231,3 +260,17 @@ reverseRepeatedFields fields x0
     reverseListField x (FieldDescriptor _ _ (RepeatedField _ f))
         = over f reverse x
     reverseListField x _ = x
+
+newtype Registry = Registry (Map.Map T.Text SomeMessageType)
```
Comment (judah) [inline proto-lens/src/Data/ProtoLens/Message.hs]: Let's add Haddocks for these functions, maybe:

    -- | A set of known message types.  Can help encode/decode protobufs containing
    -- @Data.ProtoLens.Any@ values in a more human-readable text format.
    -- 
    -- Registries can be combined using their 'Monoid' instance. 
    newtype Registry = ...

    -- | Build a 'Registry' containing a single proto type.
    register :: ...

    -- | Look up a message type by name (e.g., @"google.protobuf.FloatValue"@).
    -- The URL corresponds to the field @google.protobuf.Any.type_url@.
    lookupRegistered :: ...

File: proto-lens/src/Data/ProtoLens/TextFormat.hs [hunk 2]
```diff
@@ -203,37 +236,55 @@ modifyField (RepeatedField _ f) value = over f (value :)
...
-makeValue EnumField (Parser.EnumValue x) =
+makeValue _ EnumField (Parser.EnumValue x) =
     maybe (Left $ "Unrecognized enum value " ++ show x) Right
         (readEnum x)
-makeValue MessageField (Parser.MessageValue x) = buildMessage x
-makeValue GroupField (Parser.MessageValue x) = buildMessage x
-makeValue f val = Left $ "Type mismatch parsing text format: " ++ show (f, val)
+makeValue reg field@MessageField val@(Parser.MessageValue typeUri x)
+    | Just AnyMessageDescriptor { anyTypeUrlLens, anyValueLens } <- matchAnyMessage field =
+        case typeUri of
```
Comment (judah) [inline proto-lens/src/Data/ProtoLens/TextFormat.hs]: Can we move these pattern matches into to the guard?  If decoding the Any fails, we should just default to displaying the raw typeUrl and value (i.e. l.282) rather than failing the decode.

Also, maybe add a test for that edge case, if it's straightforward.

File: proto-lens-tests/tests/text_format.proto [hunk 3]
```diff
@@ -14,3 +16,8 @@ message Test1 {
 message Test2 {
   optional Test1 c = 1;
 }
+
+message AnyHolder {
+  optional google.protobuf.Any thing1 = 1;
+  optional google.protobuf.Any thing2 = 2;
```
Comment (tathougies) [inline proto-lens-tests/tests/text_format.proto]: good idea!

File: proto-lens/src/Data/ProtoLens/Message.hs (hunk 1 above)
Comment (tathougies) [inline proto-lens/src/Data/ProtoLens/Message.hs]: Sure

**Comment by judah** [comment]
I re-ran the Travis build and was able to get it running.  Unfortunately it failed for a couple of reasons:

- The use of TypeApplications (see my earlier message)
- The use of pattern guards (I guess it needs an explicit `LANGUAGE PatternGuards`):
https://travis-ci.org/google/proto-lens/jobs/267708755

Can you try fixing those and rebuilding with `stack build --pedantic`?

**Comment by tathougies** [comment]
Great... will do!

File: proto-lens/src/Data/ProtoLens/TextFormat.hs (hunk 2 above)
Comment (tathougies) [inline proto-lens/src/Data/ProtoLens/TextFormat.hs]: Can you explain what you mean here? This is for decoding a message which has been explicitly given a URL. I don't think there's any other meaning here other than wanting to encode an Any?

Comment (judah) [inline proto-lens/src/Data/ProtoLens/TextFormat.hs]: Ah, sorry, I got confused.  Here's a more coherent concern: if the message is in the registry but we see a raw Any (e.g. `typeUri: ...  value: "..."`) then it still ought to decode it successfully.

Something like:

    makeValue reg MessageField val@(Parser.MessageValue Nothing x)
        = buildMessage reg x
    makeValue reg field val@(Parser.MessageValue (Just uri) x)
        | Just AnyMessageDescriptor { anyTypeUrlLens, anyValueLens } <- matchAnyMessage field
            = case lookupRegistered uri reg of
                  Nothing -> Left $ "Could not decode Any of unknown type " ++ show uri
                  Just (SomeMessageType (Proxy :: Proxy value')) -> ...

File: proto-lens/src/Data/ProtoLens/TextFormat/Parser.hs [hunk 4]
```diff
@@ -67,7 +69,7 @@ data Key = Key String  -- ^ A standard key that is just a string.
 data Value = IntValue Integer  -- ^ An integer
   | DoubleValue Double  -- ^ Any floating point number
   | ByteStringValue ByteString    -- ^ A string or bytes literal
-  | MessageValue Message  -- ^ A sub message
+  | MessageValue (Maybe StrictText.Text) Message  -- ^ A sub message
```
Comment (judah) [inline proto-lens/src/Data/ProtoLens/TextFormat/Parser.hs]: Comment on the constructor what its meaning is.

**Comment by tathougies** [comment]
Any idea on why the CI keeps failing?

File: proto-lens-tests/proto-lens-tests.cabal [hunk 5]
```diff
@@ -162,6 +162,7 @@ Test-Suite text_format_test
                , proto-lens
                , proto-lens-protoc
                , proto-lens-tests
+               , proto-lens-protobuf-types
```
Comment (judah) [inline proto-lens-tests/proto-lens-tests.cabal]: You'll need to add this to the dependencies of the *library*, not the test.  Sorry for this confusing behavior; see also the comment at l.29.

File: proto-lens/src/Data/ProtoLens/TextFormat.hs [hunk 6]
```diff
@@ -203,37 +237,52 @@ modifyField (RepeatedField _ f) value = over f (value :)
...
+makeValue reg field@MessageField (Parser.MessageValue (Just typeUri) x)
+    | Just AnyMessageDescriptor { anyTypeUrlLens, anyValueLens } <- matchAnyMessage field =
+        case lookupRegistered typeUri reg of
+          Nothing -> Left "Could not decode Any"
+          Just (SomeMessageType (Proxy :: Proxy value')) ->
+            case buildMessage reg x :: Either String value' of
+              Left err -> Left err
+              Right value' -> Right (def & anyTypeUrlLens .~ typeUri
+                                         & anyValueLens .~ encodeMessage value')
+    | typeUri == messageName (descriptor :: MessageDescriptor value) =
```
Comment (judah) [inline proto-lens/src/Data/ProtoLens/TextFormat.hs]: Remove this case.  The syntax of `[pkg.Foo]{...}` (i.e.without the `type.googleapis.com/` part) refers to proto2 "extensions" which are not yet implemented in proto-lens (#27):
https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L402

And in general the type of a proto message is different than the type of an Any that wraps it, so we should avoid confusing the two.

File: proto-lens/src/Data/ProtoLens/TextFormat.hs [hunk 7]
```diff
@@ -52,28 +59,38 @@ import qualified Data.ProtoLens.TextFormat.Parser as Parser
 
 -- | Pretty-print the given message into a human-readable form.
 pprintMessage :: Message msg => msg -> Doc
-pprintMessage = pprintMessage' descriptor
+pprintMessage = pprintMessageWithRegistry mempty
+
+-- | Pretty-print the given message into human-readable form, using the given
+-- 'Registry' to decode @google.protobuf.Any@ values.
```
Comment (judah) [inline proto-lens/src/Data/ProtoLens/TextFormat.hs]: s/decode/encode

Comment (tathougies) [reply proto-lens/src/Data/ProtoLens/TextFormat.hs]: Looking back at the old reviews, you told me to change encode to decode and decode to encode :) Can we just leave it? 

In reality, it's doing both. I could say 'to encode Any values into the text format'

Comment (judah) [reply proto-lens/src/Data/ProtoLens/TextFormat.hs]: Sure, sorry about that.  Whichever one makes more sense to you.
